File: 2015/src/main/java/io/frutsel_/aoc/ADay.java
package io.frutsel_.aoc;

import java.util.ArrayList;

public abstract class ADay {

    protected final Aoc aoc;
    protected final ArrayList<String> input;

    protected ADay(Aoc aoc) {
        this.aoc = aoc;
        this.input = aoc.file(this);
    }

    public abstract int number();

    public abstract int part1();

    public abstract int part2();
}

File: aoc-commons/src/main/kotlin/nl/sbmf21/aoc/common/ADay.kt
package nl.sbmf21.aoc.common

abstract class ADay {

    val number: Int
    val input: List<String>

    init {
        Cache[this::class.java].let { (number, input) ->
            this.number = number
            this.input = input
        }
    }

    abstract fun part1(): Any

    abstract fun part2(): Any
}